src/server/v2.0/handler/label.go

```diff
@@ -0,0 +1,218 @@
+	}
+	query.Keywords["Level"] = common.LabelLevelUser
+	query.Keywords["Scope"] = scope
+	name := lib.StringValue(params.Name)
+	if name != "" {
+		query.Keywords["name"] = &q.FuzzyMatchValue{Value: name}
+	}
+	if scope == common.LabelScopeProject {
+		pid := lib.Int64Value(params.ProjectID)
+		if pid == 0 {
+			return lAPI.SendError(ctx, errors.BadRequestError(nil).WithMessage("must with project ID when to query project labels"))
+		}
+		if err := lAPI.RequireProjectAccess(ctx, pid, rbac.ActionList, rbac.ResourceLabel); err != nil {
+			return lAPI.SendError(ctx, err)
+		}
+		query.Keywords["ProjectID"] = pid
+	}
+
+	total, err := lAPI.labelMgr.Count(ctx, query)
+	if err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+	labels, err := lAPI.labelMgr.List(ctx, query)
+	if err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+
+	var results []*models.Label
+	for _, l := range labels {
+		results = append(results, model.NewLabel(l).ToSwagger())
+	}
+
+	return operation.NewListLabelsOK().
+		WithXTotalCount(total).
+		WithLink(lAPI.Links(ctx, params.HTTPRequest.URL, total, query.PageNumber, query.PageSize).String()).
+		WithPayload(results)
+}
+
+func (lAPI *labelAPI) UpdateLabel(ctx context.Context, params operation.UpdateLabelParams) middleware.Responder {
+	labelData := &pkg_model.Label{}
+	lib.JSONCopy(labelData, params.Label)
+
+	label, err := lAPI.labelMgr.Get(ctx, params.LabelID)
+	if err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+	if label == nil || label.Deleted {
+		return lAPI.SendError(ctx, errors.New(nil).WithMessage("label %d not found", params.LabelID).WithCode(errors.NotFoundCode))
+	}
+
+	if err := lAPI.requireAccess(ctx, label, rbac.ActionUpdate); err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+
+	oldName := label.Name
+	label.Name = labelData.Name
+	label.Description = labelData.Description
+	label.Color = labelData.Color
+
+	if err := label.Valid(); err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+
+	query := q.New(q.KeyWords{"Name": label.Name, "Level": label.Level, "Scope": label.Scope, "ProjectID": label.ProjectID})
+	if label.Name != oldName {
+		labels, err := lAPI.labelMgr.List(ctx, query)
+		if err != nil {
+			return lAPI.SendError(ctx, err)
+		}
+		if len(labels) > 0 {
+			return lAPI.SendError(ctx, errors.New("conflict label").WithCode(errors.ConflictCode))
+		}
+	}
+
+	if err := lAPI.labelMgr.Update(ctx, label); err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+
+	return operation.NewUpdateLabelOK()
+}
+
+func (lAPI *labelAPI) DeleteLabel(ctx context.Context, params operation.DeleteLabelParams) middleware.Responder {
+	label, err := lAPI.labelMgr.Get(ctx, params.LabelID)
+	if err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+	if err := lAPI.requireAccess(ctx, label, rbac.ActionDelete); err != nil {
+		return lAPI.SendError(ctx, err)
+	}
+	id := label.ID
+	// TODO remove this step once chart-museum is removed.
+	if err := dao.DeleteResourceLabelByLabel(id); err != nil {
```

Comment:
I found that `DeleteResourceLabelByLabel` is just a function executing a DB query,  can we move this function to the manager/dao of the label?

Reply:
see the TODO, it will be remove after chartmuseum is removed.